crates/ruma-appservice-api/src/event/push_events.rs

```diff
@@ -2,11 +2,213 @@
+struct EduDeHelper {
+    /// The message type field
+    r#type: String,
+    content: Box<RawJsonValue>,
+}
+
+#[cfg(feature = "unstable-msc2409")]
+impl<'de> Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduDeHelper { r#type, content } = from_raw_json_value(&json)?;
+
+        Ok(match r#type.as_ref() {
+            "m.presence" => Self::Presence(from_raw_json_value(&content)?),
+            "m.receipt" => Self::Receipt(from_raw_json_value(&content)?),
+            "m.typing" => Self::Typing(from_raw_json_value(&content)?),
+            _ => Self::_Custom(from_raw_json_value(&content)?),
+        })
+    }
+}
+
+/// The content for "m.presence" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+pub struct PresenceContent {
+    /// A list of presence updates that the receiving server is likely to be interested in.
+    pub push: Vec<PresenceUpdate>,
+}
+
+#[cfg(feature = "unstable-msc2409")]
+impl PresenceContent {
+    /// Creates a new `PresenceContent`.
+    pub fn new(push: Vec<PresenceUpdate>) -> Self {
+        Self { push }
+    }
+}
+
+/// An update to the presence of a user.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+pub struct PresenceUpdate {
+    /// The user ID this presence EDU is for.
+    pub user_id: OwnedUserId,
+
+    /// The presence of the user.
+    pub presence: PresenceState,
+
+    /// An optional description to accompany the presence.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub status_msg: Option<String>,
+
+    /// The number of milliseconds that have elapsed since the user last did something.
+    pub last_active_ago: UInt,
+
+    /// Whether or not the user is currently active.
+    ///
+    /// Defaults to false.
+    #[serde(default)]
+    pub currently_active: bool,
+}
+
+#[cfg(feature = "unstable-msc2409")]
+impl PresenceUpdate {
+    /// Creates a new `PresenceUpdate` with the given `user_id`, `presence` and `last_activity`.
+    pub fn new(user_id: OwnedUserId, presence: PresenceState, last_activity: UInt) -> Self {
+        Self {
+            user_id,
+            presence,
+            last_active_ago: last_activity,
+            status_msg: None,
+            currently_active: false,
+        }
+    }
+}
+
+/// The content for "m.receipt" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+pub struct ReceiptContent {
+    /// Receipts for a particular room.
+    #[serde(flatten)]
+    pub receipts: BTreeMap<OwnedRoomId, ReceiptMap>,
```

Comment:
I think this should be a tuple struct like `ReceiptEventContent` in `ruma_common::events::receipt`. The names of the inner types can probably be copied from there as well, though it makes sense that we don't reuse the same types here since this one can't include private read receipts. Side note: Please do not copy the `ReceiptType` thing, I think we should get rid of that in the events module and do the `Receipts` type more like `ReceiptMap` here.

Reply:
I hope I understood you correctly here, should this only be done for the `ReceiptContent` type?

Reply:
Yes, looks good!